pkg/callerinfo/callerinfo.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2023 Outreach Corporation. All Rights Reserved.
+
+// Description: Provides the ability to discover the module+function who is calling you.
+
+// Package callerinfo provides the GetCallerFunction to get the name of the module and function
+// that has called you.
+package callerinfo
+
+import (
+	"fmt"
+	"runtime"
+	"runtime/debug"
+	"strings"
+	"sync"
+)
+
+type CallerInfo struct {
+	Function      string
+	File          string
+	LineNum       uint
+	Module        string
+	ModuleVersion string
+}
+
+var moduleLookupByPC = make(map[uintptr]CallerInfo)
+var moduleLookupLock = sync.RWMutex{}
+
+var buildInfo *debug.BuildInfo
+
+//nolint:gochecknoinits // Why: Because we need it and the side effects are self-contained and non-mutating
+func init() {
+	if bi, ok := debug.ReadBuildInfo(); ok {
+		buildInfo = bi
+	}
+}
+
+// GetCallerInfo returns various types of info about the caller of the function.
+// skipFrames determines how many frames above the caller to skip.  If you want to know who is calling your
+// function, you would pass in skipFrames of 1.  skipFrames of 0 will return yourself.
+func GetCallerInfo(skipFrames uint16) (CallerInfo, error) {
+	// We only care about the one stack frame above the caller, so skip at least two:
+	// 1. runtime.Callers
+	// 2. GetCallerModule
+	pc := make([]uintptr, 1)
+	skipTotal := 2 + int(skipFrames)
+	num := runtime.Callers(skipTotal, pc)
+	if num == 0 {
+		return CallerInfo{}, fmt.Errorf("no frames returned from skip %d", skipTotal)
+	}
+
+	if mod, valid := moduleLookupByPC[pc[0]]; valid {
+		// Found it in the cache
+		return mod, nil
+	}
+
+	// Not cached -- have to do the slow lookup
+	frames := runtime.CallersFrames(pc)
+	frame, _ := frames.Next()
+
+	ci := CallerInfo{
+		Function: frame.Function,
+		File:     frame.File,
+		LineNum:  uint(frame.Line),
+	}
+
+	// Attempt to back-calc module name by searching through deps
+	if buildInfo != nil {
+		for _, mod := range buildInfo.Deps {
+			if strings.HasPrefix(ci.Function, mod.Path) {
+				ci.Module = mod.Path
+				ci.ModuleVersion = mod.Version
+				break
+			}
+		}
+
+		// If we can't find it in the dep list, it must be from the main app
+		if ci.Module == "" {
```

Comment:
This seems like an unsafe assumption, maybe we could at least see if `Main.Path` is in the string?

Reply:
It's been true in all my circumstances.  the function shows up as, for example, "main.main", so there's no path prefix :(

Reply:
Lame, ok. I guess we gotta :(

Can fix it later p easily I guess

Reply:
yeah, we'll see how it looks when we go wider.  this is all heuristicy :/